man/virtnbdbackup.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH VIRTNBDBACKUP "1" "September 2026" "virtnbdbackup 2.53" "User Commands"
.SH NAME
virtnbdbackup \- backup utility for libvirt
.SH DESCRIPTION
usage: virtnbdbackup [\-h] \fB\-d\fR DOMAIN [\-l {copy,full,inc,diff,auto}]
.TP
[\-t {stream,raw}] [\-r] \fB\-o\fR OUTPUT
[\-\-output\-target OUTPUT_TARGET] [\-C CHECKPOINTDIR]
[\-\-scratchdir SCRATCHDIR] [\-S] [\-\-pause] [\-i INCLUDE]
[\-x EXCLUDE] [\-f SOCKETFILE] [\-n] [\-z [COMPRESS]]
[\-w WORKER] [\-F FREEZE_MOUNTPOINT] [\-e]
[\-\-no\-sparse\-detection] [\-T THRESHOLD] [\-U URI]
[\-\-user USER] [\-\-ssh\-user SSH_USER] [\-\-ssh\-port SSH_PORT]
[\-\-ssh\-private\-key SSH_PRIVATE_KEY] [\-\-password PASSWORD]
[\-P NBD_PORT] [\-I NBD_IP] [\-\-tls] [\-\-tls\-cert TLS_CERT]
[\-L] [\-\-quiet] [\-\-nocolor] [\-q] [\-s] [\-k] [\-p] [\-v] [\-V]
.PP
Backup libvirt/qemu virtual machines
.SS "options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.SS "General options:"
.TP
\fB\-d\fR, \fB\-\-domain\fR DOMAIN
Domain to backup
.TP
\fB\-l\fR, \fB\-\-level\fR {copy,full,inc,diff,auto}
Backup level. (default: copy)
.TP
\fB\-t\fR, \fB\-\-type\fR {stream,raw}
Output type: stream or raw. (default: stream)
.TP
\fB\-r\fR, \fB\-\-raw\fR
Include full provisioned disk images in backup. (default: False)
.TP
\fB\-o\fR, \fB\-\-output\fR OUTPUT
Output target directory
.TP
\fB\-\-output\-target\fR, \fB\-\-output\-plugin\fR OUTPUT_TARGET
Output target plugin name. Defaults to 'directory', or 'zip' when output is '\-'.
.TP
\fB\-C\fR, \fB\-\-checkpointdir\fR CHECKPOINTDIR
Persistent libvirt checkpoint storage directory
.TP
\fB\-\-scratchdir\fR SCRATCHDIR
Target dir for temporary scratch file. (default: \fI\,/var/tmp\/\fP)
.TP
\fB\-S\fR, \fB\-\-start\-domain\fR
Start virtual machine if it is offline. (default: False)
.TP
\fB\-\-pause\fR
Pause virtual machine while starting backup job. (default: False)
.TP
\fB\-i\fR, \fB\-\-include\fR INCLUDE
Backup only disk with target dev name (\fB\-i\fR vda)
.TP
\fB\-x\fR, \fB\-\-exclude\fR EXCLUDE
Exclude disk(s) with target dev name (\fB\-x\fR vda,vdb)
.TP
\fB\-f\fR, \fB\-\-socketfile\fR SOCKETFILE
Use specified file for NBD Server socket (default: \fI\,/var/tmp/virtnbdbackup.3133520\/\fP)
.TP
\fB\-n\fR, \fB\-\-noprogress\fR
Disable progress bar
.TP
\fB\-z\fR, \fB\-\-compress\fR [COMPRESS]
Compress with lz4 compression level. (default: False)
.TP
\fB\-w\fR, \fB\-\-worker\fR WORKER
Amount of concurrent workers used to backup multiple disks. (default: amount of disks)
.TP
\fB\-F\fR, \fB\-\-freeze\-mountpoint\fR FREEZE_MOUNTPOINT
If qemu agent available, freeze only filesystems on specified mountpoints within virtual machine (default: all)
.TP
\fB\-e\fR, \fB\-\-strict\fR
Change exit code if warnings occur during backup operation. (default: False)
.TP
\fB\-\-no\-sparse\-detection\fR
Skip detection of sparse ranges during incremental or differential backup. (default: False)
.TP
\fB\-T\fR, \fB\-\-threshold\fR THRESHOLD
Execute backup only if threshold is reached.
.SS "Remote Backup options:"
.TP
\fB\-U\fR, \fB\-\-uri\fR URI
Libvirt connection URI. (default: qemu:///session)
.TP
\fB\-\-user\fR USER
User to authenticate against libvirtd. (default: None)
.TP
\fB\-\-ssh\-user\fR SSH_USER
User to authenticate against remote sshd: used for remote copy of files. (default: abi)
.TP
\fB\-\-ssh\-port\fR SSH_PORT
Port to connect to remote sshd: used for remote copy of files. (default: 22)
.TP
\fB\-\-ssh\-private\-key\fR SSH_PRIVATE_KEY
Path to private key file for SSH authentication (equivalent to ssh \fB\-i\fR). (default: None)
.TP
\fB\-\-password\fR PASSWORD
Password to authenticate against libvirtd. (default: None)
.TP
\fB\-P\fR, \fB\-\-nbd\-port\fR NBD_PORT
Port used by remote NBD Service, should be unique for each started backup. (default: 10809)
.TP
\fB\-I\fR, \fB\-\-nbd\-ip\fR NBD_IP
IP used to bind remote NBD service on (default: hostname returned by libvirtd)
.TP
\fB\-\-tls\fR
Enable and use TLS for NBD connection. (default: False)
.TP
\fB\-\-tls\-cert\fR TLS_CERT
Path to TLS certificates used during offline backup and restore. (default: /etc/pki/qemu/)
.SS "Logging options:"
.TP
\fB\-L\fR, \fB\-\-syslog\fR
Additionally send log messages to syslog (default: False)
.TP
\fB\-\-quiet\fR
Disable logging to stderr (default: False)
.TP
\fB\-\-nocolor\fR
Disable colored output (default: False)
.SS "Debug options:"
.TP
\fB\-q\fR, \fB\-\-qemu\fR
Use Qemu tools to query extents.
.TP
\fB\-s\fR, \fB\-\-startonly\fR
Only initialize backup job via libvirt, do not backup any data
.TP
\fB\-k\fR, \fB\-\-killonly\fR
Kill any running block job
.TP
\fB\-p\fR, \fB\-\-printonly\fR
Quit after printing estimated checkpoint size.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Enable debug output
.TP
\fB\-V\fR, \fB\-\-version\fR
Show version and exit
.SH EXAMPLES
.IP
# full backup of domain 'webvm' with all attached disks:
.IP
virtnbdbackup \-d webvm \-l full \-o /backup/
.IP
# incremental backup:
.IP
virtnbdbackup \-d webvm \-l inc \-o /backup/
.IP
# differential backup:
.IP
virtnbdbackup \-d webvm \-l diff \-o /backup/
.IP
# full backup, exclude disk 'vda':
.IP
virtnbdbackup \-d webvm \-l full \-x vda \-o /backup/
.IP
# full backup, backup only disk 'vdb':
.IP
virtnbdbackup \-d webvm \-l full \-i vdb \-o /backup/
.IP
# full backup, compression enabled:
.IP
virtnbdbackup \-d webvm \-l full \-z \-o /backup/
.IP
# full backup, create archive:
.IP
virtnbdbackup \-d webvm \-l full \-o \- > backup.zip
.IP
# full backup of vm operating on remote libvirtd:
.IP
virtnbdbackup \-U qemu+ssh://root@remotehost/system \-\-ssh\-user root \-d webvm \-l full \-o /backup/


Changelog
Version 2.53
---------
 * Add "--json" option to virtnbdbackup, virtnbdrestore and virtnbdmap: emits
 log messages (stderr, logfile and syslog) as one JSON object per line
 instead of plain text (#213)
 * Add output plugin "null" which will silenly drop data during backup
 operation: this can be quite handy for testing complete backup operations
 without wasting disk space. Can also be used for investigating possible
 performance problems ruling out the target storage as possible bottleneck,
 leaving only the NBD backend/read operation.

Version 2.52
---------
 * virtnbdbackup:
  Some fixes for bitlocker related issues:
	- Before attempting to extract bitlocker keys, check the bitlocker status
	for each volume instead of attempting to extract the keys right away.

	If bitlocker tools are found and multiple volumes are configured, this
	would lead to an useless WARNING breaking --strict mode if one of the
	volumes does not have bitlocker enabled.

	- Return and catch exception if manage-bde.exe is missing on the system.

Version 2.51
---------
 * virtnbdbackup: extend plugin layer to also include rename function for
 partial files. Add example plugin for ftp storage backend.

Version 2.50
---------
 * virtnbdrestore:
	add --relocate-data-file option to override QCOW data-file backends per
	disk during restore.
	add --remove-skipped-disks option: if option -d is specified to process
	only specific disks, adjust virtual machine configuration and remove
	disks not specified.
 * virtnbdbackup:
	Refactor the existing directory and zip output targets into a real plugin
	system that gives users the ability to implement custom plugins for
	different backup targets.

Version 2.49
---------
 * virtnbdrestore: remove mac addresses tied to network interfaces if -c
 option is set. Mac address will be auto-generated upon define (#320)

Version 2.48
---------
 * set remoteHost in case qemu+tls is used as libvirt URI (#319)

Version 2.47
---------
 * add --ssh-private-key arg for custom SSH key authentication (#315)
 * virtnbdrestore: Fix UnboundLocalError: cannot access local variable
 'dataFilePath' exception if restore of virtual machines with data-file
 setting is attempted without using -c option (#317)

Version 2.46
---------
 * virtnbdbackup: if guest agent is active, retrieve guestinfo output via
 agent and show some basic information during backup. Add the full osinfo
 output to the backup. Can be useful for inspection regarding network
 configuration and/or configured devices.
 * virtnbdbackup: In case a virtual machine has an guest agent installed and
 an Windows operating system is detected, it is now attempted to extract
 existing bitlocker recovery keys for the encrypted volumes. (#310)

Version 2.45
---------
 * virtnbdrestore: Make restore possible if no libvirt daemon is running:
 In case define option is not set, its not required to have a running libvirt
 daemon on the system restoring the disk images. (#300)

Version 2.44
---------
 * Remote backup fails: "unsupported flags (0x2) in function qemuBackupBegin":
 If system executing virtnbdbackup ships newer libvirt library version than
 remote system, VIR_DOMAIN_BACKUP_BEGIN_PRESERVE_SHUTDOWN_DOMAIN flag was
 added even if not yet supported by the remote system. Now flags are only
 activated if remote system is running on a version supporting the flags, too.
 (#298)
 * Enhance logging: log both local and remote libvirt versions.

Version 2.43
---------
 * Small code cleanups
 * Fix got an unexpected keyword argument 'stacklevel' issue on older
 distributions (#297)

Version 2.42
---------
 * virtnbdrestore: add --compress option: if enabled, NBD server used for
 restore will make use of the qcow compression driver.  This way, restored
 data within the qcow image is compressed according to the image compression
 type setting.

 This will be beneficial for images that contained many compressed blocks, as
 the restored images will be way smaller. (#295)

Version 2.41
---------
 * Disable building rust bindings when creating venv (#293)
 * Small code cleanups
 * Use tqdm logging_redirect_tqdm to not break progress bars (#294) ￼

Version 2.40
---------
 * Add OpenSUSE 16 to build scripts
 * Use trixie for docker container
 * venv/create.sh fails to successfully build libnbd without python setuptools
 (#290)
 * Add support for VIR_DOMAIN_BACKUP_BEGIN_PRESERVE_SHUTDOWN_DOMAIN: As with
 libvirt version 11.10, this flag prevents VM shutdown while an backup job is
 still active. The VM is in that scenario reset and paused instead of
 terminated allowing the backup to finish. Once the backup finishes the VM
 process is terminated. See:

  https://libvirt.org/kbase/live_full_disk_backup.html#shutdown-of-the-guest-os-during-backup

Version 2.39
---------
 * Add fedora 42 to build scripts.
 * Gracefully handle network attached disk devices, show error message and
 hint (#286)

Version 2.38
---------
 [Francisco Javier]
 * virtnbdbackup: reuse existing sftp connections during backup. (#283)

 [Michael Ablassmeier]
 * virtnbdmap: replaying changes may fail with "[Errno 22] Invalid argument":
 the NBD device sometimes needs some time to settle, wait until it is seekable
 to the end of the block list instead of failing.
 * virtnbdbackup: During full backup, all old checkpoints are removed.
 Sometimes there might be situations where bitmaps are not existent anymore.
 Now, during full backup, checkpoint consistency is validated prior to
 removal, and checkpoints tied to  non-existent bitmaps will be removed with
 VIR_DOMAIN_CHECKPOINT_DELETE_METADATA_ONLY flags enabled, to allow flawless
 full backups. The utility will issue WARNINGS in this case.

 See discussion (#285)

Version 2.37
---------
 * Add missing nbdkit and python plugin to docker container image. (#282)

Version 2.36
---------
 * virtnbdbackup: add `--pause` option: Suspend and resume virtual machine
 while starting backup job. This can be useful for virtual machines where

virtnbdbackup.py
import argparse
import nbdhelper
import extenthandler
import sparsestream
import qemuhelper
import logging
import sys

def main():
    logging.getLogger("sh").setLevel(logging.WARNING)
    logger = logging
    logger.basicConfig(level=logging.INFO)
    parser = argparse.ArgumentParser(description='Backup')
    parser.add_argument("-t", "--type", default="stream", choices=['stream','raw'], type=str, help="Output type: stream or raw")
    parser.add_argument("-q", "--qemu", default=False, help="Use Qemu tools to query extents",
        action="store_true")
    try:
        args = parser.parse_args()
    except:
        parser.print_help()
        sys.exit(1)

    nbdClient = nbdhelper.nbdClient('sda')
    connection = nbdClient.connect()

    if(args.qemu):
        logger.info("Using qemu tools to query extents")
        extentHandler = extenthandler.ExtentHandler(qemuhelper.qemuHelper('sda'))
    else:
        logger.info("Using nbd to query extents")
        extentHandler = extenthandler.ExtentHandler(connection)

    extents = extentHandler.queryBlockStatus()
    if len(extents) < 1:
        logging.error("No extents found")
        sys.exit(1)

    thinBackupSize = sum([extent.length for extent in extents if extent.data == True])
    fullBackupSize = sum([extent.length for extent in extents])
    logger.info("Found %s extents to backup" % len(extents))
    logger.info("%s bytes imagesize" % fullBackupSize)
    logger.info("%s bytes of used data in image" % thinBackupSize)

    writer = open('sda.data','wb')
    if args.type == "raw":
        logging.info("Creating full provisioned raw backup image")
        writer.truncate(fullBackupSize)
    else:
        logging.info("Creating thin provisioned stream backup image")
        metadata = sparsestream.SparseStream().dump_metadata(
            fullBackupSize,
            thinBackupSize,
            "none",
            "sda",
            False
        )
        sparsestream.SparseStream().write_frame(writer,
            sparsestream.SparseStreamTypes().META,
            0,
            len(metadata)
        )
        writer.write(metadata)
        writer.write(sparsestream.SparseStreamTypes().TERM)

    for save in extents:
        if save.data == True:
            if args.type == "stream":
                sparsestream.SparseStream().write_frame(writer,
                    sparsestream.SparseStreamTypes().DATA,
                    save.offset,
                    save.length
                )
            if save.length >= nbdClient.maxRequestSize:
                bs = nbdClient.minRequestSize
                assert save.length % bs == 0
                offset = save.offset
                count = int(save.length/bs)
                ct = 1
                while ct <= count:
                    if args.type == "raw":
                        writer.seek(offset)
                    writer.write(connection.pread(bs, offset))
                    ct+=1
                    offset+=bs
            else:
                writer.write(connection.pread(save.length, save.offset))
            if args.type == "stream":
                writer.write(sparsestream.SparseStreamTypes().TERM)
        else:
            if args.type == "raw":
                writer.seek(save.offset)
            elif args.type == "stream":
                sparsestream.SparseStream().write_frame(writer,
                    sparsestream.SparseStreamTypes().ZERO,
                    save.offset,
                    save.length
                )
    if args.type == "stream":
        sparsestream.SparseStream().write_frame(writer, sparsestream.SparseStreamTypes().STOP, 0, 0)
        writer.close()


if __name__ == "__main__":
    main()
